working/0041 - static extension methods/lrn-static-extension-survey.md

```diff
@@ -0,0 +1,177 @@
+
+Every declaration in C# must be inside a class, even static declarations. An example static extension method declaration could be:
+
+
+```c#
+public static class IntExtensions
+{
+    public static bool IsEven(this int i)
+    {
+       return ((i % 2) == 0);
+    }
+}
+```
+
+
+This declaration introduces the `IsEven` extension method on expression with type `int`, so you can write:
+
+
+```c#
+int x = 42;
+if (x.IsEven()) Console.WriteLine("Yep!");
+```
+
+
+Extension methods can be put on any type, including basic types/structs (non-reference types).
+
+
+### Resolution
+
+An invocation like `x.IsEven()` first checks if the type of `x` declares a compatible `IsEven()` method, using all the normal coercions allowed for instance method overloading resolution. If one exists, it is called. If not, the compiler then checks if there is one or more static extension method declarations in scope which matches the receiver type. If there is more than one match, the best (most specific) match is chosen, similarly to how overloaded methods are selected. If there is still more than one option, it's a compile time error.
+
+An extension method is in scope if the class containing it is.
+
+You can still call the static function normally, as `IntExtensions.isEven(42)`.
+
+C# has overloading on signatures, so conflict only happens for declarations with the same number of arguments and overlapping parameter types.
+
+Letting an interface method take precedence over an extension method avoids introducing a conflict or change when you add an extension method when the interface method already exists (likely on a subclass of the extended class, otherwise the extension wouldn't be useful at all).
+
+That is: You can't break existing functioning code by adding an extension method (well, unless the code already uses extension methods, but at the time extension methods were introduced, it was safe).
+
+
+### Generics/Nullability
+
+Extension methods can be generic, and the receiver can be nullable if it's a value type.
+
+
+```c#
+public static bool IsLessThan<T>(this IList<T> one, T other)
+{ 
+    return Nullable.Compare(one, other) < 0; 
+}
+public static T? print<T>(this T? input) where T : struct
+{
+    String output = "null";
+    if (input != null) output = input.ToString();
+    Console.WriteLine(output);
+    return input;
+}
+public static T print<T>(this T input) 
+{
+    Console.WriteLine(input.ToString());
+    return input;
+}
+```
+
+
+Since C# implements generics by specialization for value types and it reifies the type for reference types, the type `T` here will be the actual type of the receiver.
+
+Only non-nullable value types (not reference types) can be made "nullable".
```

Comment:
C# 8.0 adds support for nullable reference types: https://devblogs.microsoft.com/dotnet/nullable-reference-types-in-csharp/